File: solutions/typescript/136-find-the-duplicate-number.ts
/**
 * Problem 136: Find the Duplicate Number (LeetCode 287)
 * Difficulty: Med
 * Language: TypeScript
 */
// Floyd's cycle detection on index mapping
function findDuplicate(nums: number[]): number {
  const step = (pos: number) => nums[pos];
  const findMeet = (slow: number, fast: number): number => {
    const s = step(slow);
    const f = step(step(fast));
    return s === f ? s : findMeet(s, f);
  };
  const findStart = (a: number, b: number): number =>
    a === b ? a : findStart(step(a), step(b));
  const meet = findMeet(0, 0);
  return findStart(0, meet);
}

File: solutions/rust/136-find-the-duplicate-number.rs
/**
 * Problem 136: Find the Duplicate Number (LeetCode 287)
 * Difficulty: Med
 * Language: Rust
 */
fn find_duplicate(nums: &[i32]) -> i32 {
    let step = |i: usize| nums[i] as usize;
    fn find_meet(s: usize, f: usize, nums: &[i32]) -> usize {
        let s2 = nums[s] as usize;
        let f2 = nums[nums[f] as usize] as usize;
        if s2 == f2 { s2 } else { find_meet(s2, f2, nums) }
    }
    fn find_start(a: usize, b: usize, nums: &[i32]) -> i32 {
        if a == b { a as i32 }
        else { find_start(nums[a] as usize, nums[b] as usize, nums) }
    }
    let meet = find_meet(0, 0, nums);
    find_start(0, meet, nums)
}
